### DISCUSSION TITLE
How do I re-use a pinned map when loading an ELF?

### DISCUSSION THREAD

**Discussion by d0u9** [cilium/ebpf#603] [Q&A] [answered]
I have a map, say Global, which is pinned to `/sys/fs/bpf/global`.

I want two different BPF programs to reference this pinned map. Is that possible?

**Comment by chenhengqi** [comment]
Yes, you can do that by specifying MapOptions.PinPath.

**Comment by d0u9** [comment]
Hi @chenhengqi, thank you for your quick response.

I have an eBpf elf file that has multiple map sections. I only want to replace one of them with a pinned map. I checked the example and saw a snippet:

https://github.com/cilium/ebpf/blob/34be6948023d86f851ba48ab4020a2d193ab12a2/examples/kprobepin/main.go#L44-L55

Seems that the `PinPath` is applied to all maps in that ELF object. Is there a way to replace a single map?

**Reply by chenhengqi** [reply]
Try this:

https://github.com/cilium/ebpf/blob/3418f341ac0460cf4164146e0c1ec6c40156b4fc/map.go#L215

**Comment by ti-mo** [comment, marked as answer]
Hi @d0u9, thanks for the question.

In order for the ELF loader to automatically pin or re-use a pinned map, the map definition needs to have its pinned flag set.

See the example below: https://github.com/cilium/ebpf/blob/34be6948023d86f851ba48ab4020a2d193ab12a2/examples/kprobepin/kprobe_pin.c#L13

If you attempt to load this ELF (containing a map definition with the 'pinned' flag) without `CollectionOptions.Maps.PinPath` set, map creation will fail with an error.

Specifying `/sys/fs/bpf` will cause a map called `global` to be created and pinned to `/sys/fs/bpf/global`. If an map at that path already exists, it will automatically open it and use that existing map in any programs loaded from the ELF. The existing pinned map can then also be interacted with using the resulting `Collection.Maps["global"]`.

**Reply by mrpre** [reply]
The normal map can be declared like 
```
struct bpf_map_def SEC("maps") xxmap {
...
}
```
I wonder is it a BTF format declare in Cilium?
```
struct {
        __uint(type, BPF_MAP_TYPE_HASH);
       ...
} xxmap SEC(".maps");
``` 
and `LIBBPF_PIN_BY_NAME` only can be used along with BTF ?

**Reply by ti-mo** [reply]
@mrpre Correct, the legacy `bpf_map_def` does not have a pinning flag, so the flag cannot be set from C. If you wish to use `bpf_map_def`, you'll have to set the `MapSpec.Pinning` field in Go code after reading the ELF.
